### PR TITLE
Adds taproot support (no signing)

### PR DESCRIPTION
**Pull request by GambolingPangolin**
## Overview 

* Express taproot output structure
* Derive external key
* Parse taproot witness data

I am looking for a better solution for x-only pubkeys if anyone has suggestions.

I took the test vectors from https://github.com/bitcoin/bips/blob/995f45211d1baac4ac34685cf09d804eb8edd078/bip-0341/wallet-test-vectors.json

### CODE REVIEW

File: src/Haskoin/Crypto/Hash.hs [hunk 1]
```diff
@@ -242,3 +249,18 @@ join512 (a, b) =
     Hash512
         . BSS.toShort
         $ BSS.fromShort (getHash256 a) `BS.append` BSS.fromShort (getHash256 b)
+
+{- | Finalize tagged hash specified in BIP340
+
+@since 0.21.1
```
Comment (GambolingPangolin) [inline src/Haskoin/Crypto/Hash.hs]: Fix this version

File: src/Haskoin/Crypto/Hash.hs [hunk 2]
```diff
@@ -242,3 +249,18 @@ join512 (a, b) =
...
+-}
+initTaggedHash ::
+    -- | Hash tag
+    ByteString ->
+    -- | Hash data
```
Comment (GambolingPangolin) [inline src/Haskoin/Crypto/Hash.hs]: Incorrect comment

File: test/Haskoin/Transaction/TaprootSpec.hs [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    mapM_ checkMASTDetails $ (taprootMAST . tspkGiven) testData
+  where
+    checkMASTDetails theMAST = do
+        -- Leaf hashes
+        (Just . getLeafHashes) theMAST @?= (spkiLeafHashes . tspkIntermediary) testData
+        -- Merkle root
+        (Just . BA.convert . mastCommitment) theMAST @?= (spkiMerkleRoot . tspkIntermediary) testData
+
+    getLeafHashes =
+        \case
```
Comment (GambolingPangolin) [inline test/Haskoin/Transaction/TaprootSpec.hs]: Move `\case` to line above

**Comment by jprupp** [comment]
I don't know much about Taproot. You may squash and merge if you are satisfied.

**Comment by jprupp** [comment]
I want to do a release today or tomorrow. I'll merge this first. You may do another pull request if you have anything else to add.

**Comment by GambolingPangolin** [comment]
@jprupp Ok sounds good.  I'm happy with the patch, myself 😃

**Pull request merged**